### PR TITLE
Implement Router Z-loss

### PR DESCRIPTION
**Pull request by josejg**
# What does this PR do?

Z-loss is an additional term that improves the stability of softmax logit inputs by penalizing large values where the float precision is reduced.

Idea was introduced by the ST-MoE Paper - https://arxiv.org/abs/2202.08906. And has been recently used in the OLMoE work - https://arxiv.org/abs/2409.02060.

# What issue(s) does this change relate to?

Alternative implementation of https://github.com/databricks/megablocks/pull/133 

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/databricks/megablocks/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
  - Unsure where to document things
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/databricks/megablocks/blob/dev/CONTRIBUTING.md#running-tests))
  - Added a `test_moe_forward_backward_with_zloss` and `test_moe_forward_backward_with_zloss`
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/databricks/megablocks/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mihir-db — APPROVED** [review]
LGTM!
